### PR TITLE
[2][cleanup][gcs] Cleanup GCS client options.

### PR DESCRIPTION
**Pull request by fishbone**
## Why are these changes needed?
This PR cleanup GCS client options.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by ericl — APPROVED** [review]

**Comment by suquark** [comment]
there are several C++ CI test errors

File: src/ray/gcs/gcs_client/gcs_client.cc [hunk 1]
```diff
@@ -89,42 +85,13 @@ GcsClient::GcsClient(
...
-                                            options_.redis_port_,
-                                            options_.password_,
-                                            /*enable_sharding_conn=*/false,
-                                            options_.enable_sync_conn_,
-                                            options_.enable_async_conn_,
-                                            options_.enable_subscribe_conn_);
-    redis_client_ = std::make_shared<RedisClient>(redis_client_options);
-    RAY_CHECK_OK(redis_client_->Connect(io_service));
-  } else {
-    RAY_CHECK(::RayConfig::instance().gcs_grpc_based_pubsub())
```
Comment (kfstorm) [inline src/ray/gcs/gcs_client/gcs_client.cc:111]: Shall we delete `gcs_grpc_based_pubsub` in `ray_config_def.h`?

File: src/ray/gcs/gcs_client/gcs_client.h [hunk 2]
```diff
@@ -77,18 +55,6 @@ class GcsClientOptions {
...
-  std::string redis_ip_;
-  int redis_port_ = 0;
-
-  // Password of GCS server.
-  std::string password_;
-
-  // Whether to enable connection for contexts.
-  bool enable_sync_conn_{true};
-  bool enable_async_conn_{true};
-  bool enable_subscribe_conn_{true};
```
Comment (kfstorm) [inline src/ray/gcs/gcs_client/gcs_client.h:91]: We can delete these options in `RedisClientOptions` as well.

**Review by kfstorm — COMMENTED** [review]
Is it possible to remove hiredis dependency in Bazel for Raylet and CoreWorker now?

File: src/ray/gcs/gcs_client/gcs_client.cc (hunk 1 above)
Comment (iycheng) [reply src/ray/gcs/gcs_client/gcs_client.cc:111]: @kfstorm  it's in the next PR  (https://github.com/ray-project/ray/pull/23520)

**Comment by iycheng** [comment]
> Is it possible to remove hiredis dependency in Bazel for Raylet and CoreWorker now?

Good point. Let me first clean the code and then remove the dependences in bazel.

**Review by mwtian — APPROVED** [review]

File: src/ray/gcs/gcs_client/gcs_client.h (hunk 2 above)
Comment (iycheng) [reply src/ray/gcs/gcs_client/gcs_client.h:91]: I'll update it in the next PR.

**Pull request merged**